### ISSUE TITLE
Location List Doesn't Close After All Errors Have Been Addressed

### ISSUE THREAD

**Issue by jfelchner** [w0rp/ale#2460]
## Information

**VIM version**

```
NVIM v0.3.4
Build type: Release
```

Operating System: macOS 10.14

## What went wrong

I have an error in my file, which shows up in the quickfix list.  I fix the error in my file.  ALE does not auto-close the location list once the error has been corrected.

Additionally if I have an error, then _close_ the location list, ALE opens the location list again, shows the error, runs the checks, then leaves me with an empty location list.

a) The location list should not reopen until ALE is sure that there are errors remaining in the file after the checks run.

b) We need a way to auto-close the location list if there are no more errors after checks.  The `QuitPre` autocommand only works if you're closing the buffer, which I am not.

## Reproducing the bug

See above

### :ALEInfo

https://gist.github.com/jfelchner/77a37bde7fdff2275f1c5bf7252b30e8

**Comment by w0rp** [comment]
It's supposed to do what you describe. Let me know which linters you were using, as the `:ALEInfo` output there seems to be for another buffer where no linters are enabled.

**Comment by jfelchner** [comment]
@w0rp I updated the gist

**Comment by CorbanR** [comment]
I am experiencing the same issue

```NVIM v0.4.0-1240-ga3d8cd3f6
Build type: Release```

**:ALEInfo**
https://gist.github.com/CorbanR/0ed4833d9e3dd836a904a77559a629b0

**Relevant .vimrc config**
https://gist.github.com/CorbanR/2a57a7e0fa7beb1ef33506cff575c366

<img width="1675" alt="Screen Shot 2019-04-28 at 2 03 51 PM" src="https://user-images.githubusercontent.com/1918683/56869636-0630a780-69c0-11e9-9594-80276f3f7ac8.png">

<img width="1677" alt="Screen Shot 2019-04-28 at 2 04 19 PM" src="https://user-images.githubusercontent.com/1918683/56869643-1779b400-69c0-11e9-8068-b4fd4abdb325.png">

**Comment by w0rp** [comment]
This was probably caused by #2415. `git pull` and try it again. If the issue persists, debug it and submit a patch for it. I can't repeat the bug myself.

**Comment by CorbanR** [comment]
Thanks for creating this plugin w0rp, Ale is srsly amazing.
Still appears to be an issue. Again, it seems tied to `solargraph` being enabled as a linter. 

When `solargraph` is not loaded as a linter it does NOT seem to be a problem.

```
let b:ale_linters = {
\    'ruby': ['brakeman', 'rails_best_practices', 'reek', 'rubocop', 'ruby', 'standardrb']
\}
```

**Comment by w0rp** [comment]
I have fixed this now. The problem was specific to only Language Server Protocol. The problem is that it's impossible to tell if a language server will respond to a file changing or being saved with diagnostics, so ALE can't keep track of if results are pending from language servers, as they may never arrive. ALE does this reliably for `tsserver`, which requires an explicit message to request diagnostics, and which always returns a response for that message. ALE was using that information for determining if the location list window should be closed, to prevent it from popping open and closed too often. (Which for regular command line linters is very annoying.)

I have created an issue for the protocol here. https://github.com/Microsoft/language-server-protocol/issues/737 I think the protocol should be updated to support a mode which means that diagnostics are only sent to clients that explicitly request them, for a number of reasons.
